src/neo/Ledger/Blockchain.cs

```diff
@@ -415,28 +415,36 @@ private void Persist(Block block)
                     Context.System.EventStream.Publish(application_executed);
                     all_application_executed.Add(application_executed);
                 }
-                foreach (IPersistencePlugin plugin in Plugin.PersistencePlugins)
-                    plugin.OnPersist(system, block, snapshot, all_application_executed);
-                snapshot.Commit();
-                List<Exception> commitExceptions = null;
                 foreach (IPersistencePlugin plugin in Plugin.PersistencePlugins)
                 {
-                    try
+                    if (plugin.CurrentHeight < block.Index)
                     {
-                        plugin.OnCommit(system, block, snapshot);
+                        plugin.OnPersist(system, block, snapshot, all_application_executed);
```

Comment:
Maybe `OnPersist` and `OnCommit` could be merged then? Previously they were done at different stages of processing, but now they're very similar and probably this separation is not really needed. It will change plugin interface though (but I doubt we have a lot of third-party plugins now).

Reply:
This is introduced in #568:

> If there are multiple persistence plugins loaded and one of them throws an exception, then none of them should actually commit anything to a persistent store. To facilitate this behavior, I added an OnCommit method to be called after all plugins have had a chance to process results from OnPersist and throw if something was wrong.

Reply:
Yes, we should merge it

Reply:
I think we'd better keep this strategy to make sure there is no dirty data stored in plugins once some error occurs

Reply:
But it's better to revert if there is an error

Reply:
Once we merge these 2 functions, it will not be able to work correctly for such scenarios: i.e. there are 2 IPersistencePlugins installed. The first one persists without error but the second one throws some exception caused by dirty data. We can restart nodes again to re-persist data for neo-node and IPersistencePlugins, but the first plugin is already "polluted", I'm afraid the node can only be re-syncronized from 0 height again...

Reply:
If there are two plugins, and one throw an exception the problem will be in the plugin, I still thinking that is better to merge it. What do you think @erikzhang ?

Reply:
Ping

Reply:
@erikzhang Could you provide opinion upob this question?